docs/man/memcached.3
.TH "MEMCACHED" "3" "September 16, 2012" "1.0.10" "libmemcached"
.SH NAME
memcached \- libmemcached Documentation
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.\" Man page generated from reStructeredText.
.
.SH SYNOPSIS
.sp
#include <libmemcached\-1.0/memcached.h>
.INDENT 0.0
.TP
.B LIBMEMCACHED
.UNINDENT
.INDENT 0.0
.TP
.B memcached_st *memcached(const char\fI\ *string\fP, size_t\fI\ string_length\fP)
.UNINDENT
.INDENT 0.0
.TP
.B memcached_return_t libmemcached_check_configuration(const char\fI\ *option_string\fP, size_t\fI\ length\fP, char\fI\ *error_buffer\fP, size_t\fI\ error_buffer_size\fP)
.UNINDENT
.sp
Compile and link with \-lmemcached
.SH DESCRIPTION
.sp
Libmemcached implements a custom language for configuring and modifying
servers. By passing in an option string you can generate a \fBmemcached_st\fP object
that you can use in your application directly.
.INDENT 0.0
.TP
.B \-\-SERVER=<servername>:<optional_port>/?<optional_weight>
.UNINDENT
.sp
Provide a servername to be used by the client. Providing a weight will cause weighting to occur with all hosts with each server getting a default weight of 1.
.INDENT 0.0
.TP
.B \-\-SOCKET="<filepath>/?<optional_weight>"
.UNINDENT
.sp
Provide a filepath to a UNIX socket file. Providing a weight will cause weighting to occur with all hosts with each server getting a default weight of 1.
.INDENT 0.0
.TP
.B \-\-VERIFY\-KEY
.UNINDENT
.sp
Verify that keys that are being used fit within the design of the protocol being used.
.INDENT 0.0
.TP
.B \-\-REMOVE_FAILED_SERVERS
.UNINDENT
.sp
Enable the behavior \fBMEMCACHED_BEHAVIOR_REMOVE_FAILED_SERVERS\fP.
.INDENT 0.0
.TP
.B \-\-BINARY\-PROTOCOL
.UNINDENT
.sp
Force all connections to use the binary protocol.
.INDENT 0.0
.TP
.B \-\-BUFFER\-REQUESTS
.UNINDENT
.sp
Please see \fBMEMCACHED_BEHAVIOR_BUFFER_REQUESTS\fP.
.INDENT 0.0
.TP
.B \-\-CONFIGURE\-FILE=
.UNINDENT
.sp
Provide a configuration file to be used to load requests. Beware that by using a configuration file libmemcached will reset memcached_st based on information only contained in the file.
.INDENT 0.0
.TP
.B \-\-CONNECT\-TIMEOUT=
.UNINDENT
.sp
Please see \fBMEMCACHED_BEHAVIOR_CONNECT_TIMEOUT\fP.
.INDENT 0.0
.TP
.B \-\-DISTRIBUTION=
.UNINDENT
.sp
Set the distribution model used by the client.  See :manpage:\(ga\(ga for more details.
.INDENT 0.0
.TP
.B \-\-HASH=
.UNINDENT
.sp
Set the hashing alogrthm used for placing keys on servers.
.INDENT 0.0
.TP
.B \-\-HASH\-WITH\-NAMESPACE
.UNINDENT
.sp
When enabled the prefix key will be added to the key when determining which
server to store the data in.
.INDENT 0.0
.TP
.B \-\-NOREPLY
.UNINDENT
.sp
Enable "no reply" for all calls that support this. It is highly recommended
that you use this option with the binary protocol only.
.INDENT 0.0
.TP
.B \-\-NUMBER\-OF\-REPLICAS=
.UNINDENT
.sp
Set the nummber of servers that keys will be replicated to.
.INDENT 0.0
.TP
.B \-\-RANDOMIZE\-REPLICA\-READ
.UNINDENT
.sp
Select randomly the server within the replication pool to read from.
.INDENT 0.0
.TP
.B \-\-SORT\-HOSTS
.UNINDENT
.sp
When adding new servers always calculate their distribution based on sorted naming order.
.INDENT 0.0
.TP
.B \-\-SUPPORT\-CAS
.UNINDENT
.sp
See \fImemcached_behavior_set(3)\fP for \fBMEMCACHED_BEHAVIOR_SUPPORT_CAS\fP
.INDENT 0.0
.TP
.B \-\-USE\-UDP
.UNINDENT
.sp
See \fImemcached_behavior_set(3)\fP for \fBMEMCACHED_BEHAVIOR_USE_UDP\fP
.INDENT 0.0
.TP
.B \-\-NAMESPACE=
.UNINDENT
.sp
A namespace is a container that provides context for keys, only other
requests that know the namespace can access these values. This is
accomplished by prepending the namespace value to all keys.
.SS Mecached Pool Options:
.INDENT 0.0
.TP
.B \-\-POOL\-MIN
.UNINDENT
.sp
Initial size of pool.
.INDENT 0.0
.TP
.B \-\-POOL\-MAX
.UNINDENT
.sp
Maximize size of the pool.
.SS I/O Options:
.INDENT 0.0
.TP
.B \-\-TCP\-NODELAY
.UNINDENT
.sp
See \fImemcached_behavior_set(3)\fP for MEMCACHED_BEHAVIOR_TCP_NODELAY
.INDENT 0.0
.TP
.B \-\-TCP\-KEEPALIVE
.UNINDENT
.sp
See \fImemcached_behavior_set(3)\fP for MEMCACHED_BEHAVIOR_TCP_KEEPALIVE
.INDENT 0.0
.TP
.B \-\-RETRY\-TIMEOUT=
.UNINDENT
.sp
See \fImemcached_behavior_set(3)\fP for MEMCACHED_BEHAVIOR_RETRY_TIMEOUT
.INDENT 0.0
.TP
.B \-\-SERVER\-FAILURE\-LIMIT=
.UNINDENT
.sp
See \fImemcached_behavior_set(3)\fP for MEMCACHED_BEHAVIOR_SERVER_FAILURE_LIMIT
.INDENT 0.0
.TP
.B \-\-SND\-TIMEOUT=
.UNINDENT
.sp
See \fImemcached_behavior_set(3)\fP for MEMCACHED_BEHAVIOR_SND_TIMEOUT
.INDENT 0.0
.TP
.B \-\-SOCKET\-RECV\-SIZE=
.UNINDENT
.sp
See \fImemcached_behavior_set(3)\fP for MEMCACHED_BEHAVIOR_SOCKET_RECV_SIZE
.INDENT 0.0
.TP
.B \-\-SOCKET\-SEND\-SIZE=
.UNINDENT
.sp
See \fImemcached_behavior_set(3)\fP for MEMCACHED_BEHAVIOR_SOCKET_SEND_SIZE
.INDENT 0.0
.TP
.B \-\-POLL\-TIMEOUT=
.UNINDENT
.sp
That sets the value of the timeout used by :manpage: \fIpoll()\fP.
.INDENT 0.0
.TP
.B \-\-IO\-BYTES\-WATERMARK=
.UNINDENT
.INDENT 0.0
.TP
.B \-\-IO\-KEY\-PREFETCH=
.UNINDENT
.INDENT 0.0
.TP
.B \-\-IO\-MSG\-WATERMARK=
.UNINDENT
.INDENT 0.0
.TP
.B \-\-TCP\-KEEPIDLE
.UNINDENT
.INDENT 0.0
.TP
.B \-\-RCV\-TIMEOUT=
.UNINDENT
.SS Other:
.INDENT 0.0
.TP
.B INCLUDE
.UNINDENT
.sp
Include a file in configuration. Unlike \-\-CONFIGURE\-FILE= this will not reset memcached_st
.INDENT 0.0
.TP
.B RESET
.UNINDENT
.sp
Reset memcached_st and continue to process.
.INDENT 0.0
.TP
.B END
.UNINDENT
.sp
End configutation processing.
.INDENT 0.0
.TP
.B ERROR
.UNINDENT
.sp
End configutation processing and throw an error.
.SH RETURN
.sp
\fI\%memcached()\fP returns a pointer to the memcached_st that was
created (or initialized).  On an allocation failure, it returns NULL.
.SH EXAMPLE
.sp
.nf
.ft C
const char *config_string= "\-\-SERVER=host10.example.com \-\-SERVER=host11.example.com \-\-SERVER=host10.example.com"
memcached_st *memc= memcached(config_string, strlen(config_string);
{
 ...
}
memcached_free(memc);
.ft P
.fi
.SH HOME
.sp
To find out more information please check:
\fI\%http://libmemcached.org/\fP
.SH SEE ALSO
.sp
\fImemcached(1)\fP \fIlibmemcached(3)\fP \fImemcached_strerror(3)\fP
.SH AUTHOR
Brian Aker
.SH COPYRIGHT
2011-2012, Brian Aker DataDifferential, http://datadifferential.com/
.\" Generated by docutils manpage writer.
.\" 
.


ChangeLog
1.0.11 Sun Sep 16 20:32:13 EDT 2012
* Removed custom version of memcached.
* Updated hardening rules.
* Fixed a case where the return error from a socket connection differred from that of a TCP/IP socket.

1.0.10 Sun Jul 29 21:50:15 PDT 2012
* --disable-assert has been removed from configure, and --enable-assert has been added in its place.
* Compiling fixes for Clang on OSX Mountain Lion.

1.0.9 Wed Jul  4 22:46:19 EDT 2012
* Faster close on socket.
* Instance allocation is now seperated from server interface. This should allow for a better preservation of ABI compliance from now on.
* Fix close on exec bug.
* Numerous other bug fixes.

1.0.8 Tue May 22 15:06:04 EDT 2012
* Added support for setting options via ENV variable LIBMEMCACHED
* Fix corner case on last used result.

1.0.7 Sat Apr 28 00:48:29 PDT 2012
* Add API call for exist calls.
* Update all license files to be BSD.

1.0.6 Sat Apr  7 18:26:49 PDT 2012
* Fixes for gcc 4.7, lp:961812
* Fix for restart issue that happens under testing.
* Fix for lp:962815.
* Support for transparent AES encryption.

1.0.5 Tue Mar 13 22:56:47 PDT 2012
* Fixes for OSX.
* Version is now parsed directly in the parser, which makes buffered operations now work with it..
* memstat has been extended so that it can be used to find the version of the server.
* Update documentation.
* Fixes for compile issues on Debian and Ubuntu

1.0.4 Thu Jan 26 22:33:54 PST 2012
* Fix for memcached_dump().
* Additional testing for memcached_stat_execute().

1.0.3 Sun Jan  8 18:11:36 PST 2012
* Increased size of sort buffer used during Ketama.
* Added support for new behavior to handle dead servers.
* Overall haul of UDP IO.
* Fixed C compile issue with memcached_exist()
* Numerous bug fixes.
* Clang support for OSX.
* All commands now using vector send support.

1.0.2 Mon Oct 24 08:00:48 PDT 2011
* Dropped libmemcached/memcached_util.h (undocumented header file)
* Added memcached_touch() and memcached_touch_by_key()
* UDP support restructured to toggle on a complete memcached_st structure.

0.53 Mon Sep 26 20:50:33 PDT 2011
  * Fix for FreeBSD/OpenBSD and -lm
  * Added memcached_exist()
  * Fix for memory when using config test.
  * CLI gained --quiet

0.52 Sun Sep 11 22:16:08 PDT 2011
  * Build fixes for Ubuntu/Suse.
  * Fixes for OSX Lion.
  * Bug fix for looping back through dns lookups under certain failures.
  * Fixes related to dead server failures.

0.51 Thu Jul 21 11:08:58 PDT 2011
  * memcached_callback_set() now takes its data argument as const
  * Update to tests.
  * Fix in parser for port number.

0.50 Mon Jun 20 10:36:57 PDT 2011
 * Updates to C++ interface 
 * Custom free allocators need to now check for value before calling free.
 * memcached_fetch_result() now uses the internal result when available (about 25 to 50% faster).
 * Fix for stats structure.
 * Updates to documentation.
 * memcached_fetch_result() now uses the internal result when available (about 25 to 50% faster).

0.49 Thu Apr 14 08:43:37 PDT 2011
  * Fix calls to auto methods so that if value is not passed in nothing bad happens.
  * New parser calls for generating memcached_st objects.
  * New error system.
  * New flow control for messages means faster get/set calls.
  * Added new documentation system.
  * A behavior change has been now made that if you specify a weight for any server, we enable the weight flag and do weight balancing.  
  * Added MEMCACHED_BEHAVIOR_REMOVE_FAILED_SERVERS to simplify the setting of AUTO REJECT for servers.

0.48 Tue Mar 15 23:05:18 PDT 2011
  * Fix memory leak in server parse.
  * Move test framework out to be its own library (easier to work with Gearman).


0.47 Wed Feb 23 18:46:48 PST 2011
  * Additional fixes for OpenBSD.
  * Bug fix 677609, 456080.
  * SIGPIPE fix for Linux send(). 
  * memcapable can now test ascii or binary based on flags.
  * Additional build fixes for SASL.


0.46 Mon Feb 14 10:28:01 PST 2011
  * Fixes a number of corner case bugs.
  * Fixes related to OpenBSD.
  * Better testing for protocol version.
  * Removes special case infinite wait on blocking setup.

0.45 Tue Feb  8 16:02:06 PST 2011
  * Add support for systemtap

0.44 Wed Sep 22 21:57:57 PDT 2010
  * Windows bug fixes.
  * Hudson port support in test harness.
  * Improved portability of test hanrness.
  * SASL fixes.

0.43 Wed Jul 28 16:29:47 PDT 2010
  * Added --args to memstat so that a greater range of values can be returned.
  * Prelimanary support for Windows.
  * memcached_stat_execute() merged. 

0.42 Tue Jul  6 12:29:50 PDT 2010
  * Mistake in libtool caused issue with library version

0.41 Wed Jun 30 13:15:19 PDT 2010
  * Added --file for memcat.
  * Added limemcached_ping() to libmemcached_util
  * Bugfix for some cases where connect would have issues with timeout.
  * Wrong value for errno given as error on an IO failure inside of poll.
  * Bug fix for issue where multiple interfaces with bad DNS were not being caught.

0.40 Thu Apr 22 19:01:25 PDT 2010
  * Placed retry logic in for busted resolvers
  * Add an ignore for SIGPIPE to solve OSX issues.
  * A couple of fixed for memcached_light server.
  * Updated to debug mode to track io_wait

0.39 Tue Apr  6 12:35:13 PDT 2010
  * Add support for prefix keys to binary protocol.
  * Remove the undocumented call memcached_server_remove().
  * The undocumented call memcached_server_by_key() now returns const.
  * memcached_server_error_reset() has been deprecated.
  * memcached_server_list() has been deprecated. Use memcached_server_cursor() to walk the servers found in a memcached_st() structure.
  * memcached_verbosity() can now be run concurrently with other operations.
  * SASL support.
  * Fixes memory leak found in EJECT HOSTS.

0.38 Wed Feb 10 09:40:33 PST 2010
  * C++ interface for libhashkit.
  * Modified memcached_set_memory_allocators() so that it requires a context pointer.
  * memcached_clone() now runs 5 times faster.
  * Functions used for callbacks are now given const memcached_st.

libtest/memcached.cc
/*  vim:expandtab:shiftwidth=2:tabstop=2:smarttab:
 *
 *  Data Differential YATL (i.e. libtest)  library
 *
 *  Copyright (C) 2012 Data Differential, http://datadifferential.com/
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions are
 *  met:
 *
 *      * Redistributions of source code must retain the above copyright
 *  notice, this list of conditions and the following disclaimer.
 *
 *      * Redistributions in binary form must reproduce the above
 *  copyright notice, this list of conditions and the following disclaimer
 *  in the documentation and/or other materials provided with the
 *  distribution.
 *
 *      * The names of its contributors may not be used to endorse or
 *  promote products derived from this software without specific prior
 *  written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 *  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 *  OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 *  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 *  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 *  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 *  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 *  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 *  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 */

#include <config.h>
#include <libtest/common.h>

#include <libmemcached-1.0/memcached.h>
#include <libmemcachedutil-1.0/util.h>

using namespace libtest;

#include <cassert>
#include <cerrno>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <iostream>
#include <signal.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>

#include <libtest/server.h>
#include <libtest/wait.h>

#include <libtest/memcached.h>

#ifndef __INTEL_COMPILER
#pragma GCC diagnostic ignored "-Wold-style-cast"
#endif

using namespace libtest;

namespace {
  bool is_memcached_libtool()
  {
    if (MEMCACHED_BINARY and strcmp(MEMCACHED_BINARY, "memcached/memcached") == 0) 
    {
      return true;
    }

    return false;
  }
}

class Memcached : public libtest::Server
{
  std::string _username;
  std::string _password;

public:
  Memcached(const std::string& host_arg,
            const in_port_t port_arg,
            const bool is_socket_arg,
            const std::string& username_arg,
            const std::string& password_arg) :
    libtest::Server(host_arg, port_arg, 
                    MEMCACHED_BINARY, is_memcached_libtool(), is_socket_arg),
    _username(username_arg),
    _password(password_arg)
  { }

  Memcached(const std::string& host_arg, const in_port_t port_arg, const bool is_socket_arg) :
    libtest::Server(host_arg, port_arg,
                    MEMCACHED_BINARY, is_memcached_libtool(), is_socket_arg)
  {
    set_pid_file();
  }

  virtual const char *sasl() const
  {
    return NULL;
  }

  const std::string& password() const
  {
    return _password;
  }

  const std::string& username() const
  {
    return _username;
  }

  bool wait_for_pidfile() const
  {
    Wait wait(pid(), 4);

    return wait.successful();
  }

  bool ping()
  {
#if 0
    // Memcached is slow to start, so we need to do this
    if (pid_file().empty() == false)
    {
      if (wait_for_pidfile() == false)
      {
        Error << "Pidfile was not found:" << pid_file() << " :" << running();
        return -1;
      }
    }
#endif

    memcached_return_t rc;
    bool ret;

    if (has_socket())
    {
      ret= libmemcached_util_ping(socket().c_str(), 0, &rc);
    }
    else
    {
      ret= libmemcached_util_ping(hostname().c_str(), port(), &rc);
    }

    if (memcached_failed(rc) or ret == false)
    {
      error(memcached_strerror(NULL, rc));
    }

    return ret;
  }

  const char *name()
  {
    return "memcached";
  };

  const char *executable()
  {
    return MEMCACHED_BINARY;
  }

  bool is_libtool()
  {
    return is_memcached_libtool();
  }

  virtual void pid_file_option(Application& app, const std::string& arg)
  {
    if (arg.empty() == false)
    {
      app.add_option("-P", arg);
    }
  }

  const char *socket_file_option() const
  {
    return "-s ";
  }

  virtual void port_option(Application& app, in_port_t arg)
  {
    char buffer[30];
    snprintf(buffer, sizeof(buffer), "%d", int(arg));
    app.add_option("-p", buffer); 
  }

  bool has_port_option() const
  {
    return true;
  }

  bool has_socket_file_option() const
  {
    return has_socket();
  }

  void socket_file_option(Application& app, const std::string& socket_arg)
  {
    if (socket_arg.empty() == false)
    {
      app.add_option("-s", socket_arg);
    }
  }

  bool broken_socket_cleanup()
  {
    return true;
  }

  // Memcached's pidfile is broken
  bool broken_pid_file()
  {
    return true;
  }

  bool build(size_t argc, const char *argv[]);
};

class MemcachedLight : public libtest::Server
{

public:
  MemcachedLight(const std::string& host_arg, const in_port_t port_arg) :
    libtest::Server(host_arg, port_arg, MEMCACHED_LIGHT_BINARY, true)
  {
    set_pid_file();
  }

  bool ping()
  {
    // Memcached is slow to start, so we need to do this
    if (not pid_file().empty())
    {
      if (not wait_for_pidfile())
      {
        Error << "Pidfile was not found:" << pid_file();
        return false;
      }
    }

    std::stringstream error_message;
    pid_t local_pid= get_pid_from_file(pid_file(), error_message);
    if (local_pid > 0)
    {
      if (::kill(local_pid, 0) == 0)
      {
        return true;
      }
    }

    return false;
  }

  const char *name()
  {
    return "memcached_light";
  };

  const char *executable()
  {
    return MEMCACHED_LIGHT_BINARY;
  }

  virtual void port_option(Application& app, in_port_t arg)
  {
    char buffer[1024];
    snprintf(buffer, sizeof(buffer), "--port=%d", int(arg));
    app.add_option(buffer);
  }

  bool has_port_option() const
  {
    return true;
  }

  bool is_libtool()
  {
    return true;
  }

  void log_file_option(Application& app, const std::string& arg)
  {
    if (arg.empty() == false)
    {
      std::string buffer("--log-file=");
      buffer+= arg;
      app.add_option("--verbose");
      app.add_option(buffer);
    }
  }

  bool has_log_file_option() const
  {
    return true;
  }

  bool build(size_t argc, const char *argv[]);
};

class MemcachedSaSL : public Memcached
{
public:
  MemcachedSaSL(const std::string& host_arg,
                const in_port_t port_arg, 
                const bool is_socket_arg, 
                const std::string& username_arg, 
                const std::string &password_arg) :
    Memcached(host_arg, port_arg, is_socket_arg, username_arg, password_arg)
  { }

  const char *name()
  {
    return "memcached-sasl";
  };

  const char *sasl() const
  {
    return " -S -B binary ";
  }

  const char *executable()
  {
    return MEMCACHED_SASL_BINARY;
  }

  bool ping()
  {
    memcached_return_t rc;
    bool ret;

    if (has_socket())
    {
      ret= libmemcached_util_ping2(socket().c_str(), 0, username().c_str(), password().c_str(), &rc);
    }
    else
    {
      ret= libmemcached_util_ping2(hostname().c_str(), port(), username().c_str(), password().c_str(), &rc);
    }

    if (memcached_failed(rc) or ret == false)
    {
      error(memcached_strerror(NULL, rc));
    }

    return ret;
  }

};


#include <sstream>

bool Memcached::build(size_t argc, const char *argv[])
{
  if (getuid() == 0 or geteuid() == 0)
  {
    add_option("-u", "root");
  }

  add_option("-l", "localhost");
#ifndef TARGET_OS_OSX
  add_option("-m", "128");
  add_option("-M");
#endif

  if (sasl())
  {
    add_option(sasl());
  }

  for (size_t x= 0 ; x < argc ; x++)
  {
    add_option(argv[x]);
  }

  return true;
}

bool MemcachedLight::build(size_t argc, const char *argv[])
{
  for (size_t x= 0 ; x < argc ; x++)
  {
    add_option(argv[x]);
  }

  return true;
}

namespace libtest {

libtest::Server *build_memcached(const std::string& hostname, const in_port_t try_port)
{
  return new Memcached(hostname, try_port, false);
}

libtest::Server *build_memcached_socket(const std::string& socket_file, const in_port_t try_port)
{
  return new Memcached(socket_file, try_port, true);
}

libtest::Server *build_memcached_light(const std::string& hostname, const in_port_t try_port)
{
  return new MemcachedLight(hostname, try_port);
}


libtest::Server *build_memcached_sasl(const std::string& hostname, const in_port_t try_port, const std::string& username, const std::string &password)
{
  if (username.empty())
  {
    return new MemcachedSaSL(hostname, try_port, false,  "memcached", "memcached");
  }

  return new MemcachedSaSL(hostname, try_port, false,  username, password);
}

libtest::Server *build_memcached_sasl_socket(const std::string& socket_file, const in_port_t try_port, const std::string& username, const std::string &password)
{
  if (username.empty())
  {
    return new MemcachedSaSL(socket_file, try_port, true, "memcached", "memcached");
  }

  return new MemcachedSaSL(socket_file, try_port, true, username, password);
}

}

